### ISSUE TITLE
Unit Tests

### ISSUE THREAD

**Issue by Seanny123** [nengo/nengo_gui#471]
Making unit tests with Selenium is really easy (good docs [here](https://selenium-python.readthedocs.org/en/latest/index.html) and [here](https://selenium.googlecode.com/svn/trunk/docs/api/py/index.html)). I'm not saying we should test every feature, but a few quick sanity tests would be nice.

Off the top of my head:
- to see if we've broken any graphs
- check if iPython hasn't exploded
- check if the basic network generation (does entering new code generate new items?) is functional.

But before I get ahead of myself listing a bunch of things to test, we need to agree on how we're going to do this. @pblouw and @arvoelke you were apparently working on this before the summer school? Did you use Selenium?

**Comment by pblouw** [comment]
We had a number of tests running in Selenium covering basic things like
making graphs, starting and stopping a simulation, changing input values,
etc. The tests are almost certainly broken now due to changes to the gui,
but its been on my todo list to update them and get them working properly
again. Suggestions for a list of features to test would be welcome.

**Comment by Seanny123** [comment]
Where can I find these old tests?

**Comment by pblouw** [comment]
I don't think I pushed them to github, so you'll have to come talk to me.

**Comment by Seanny123** [comment]
Having gotten the tests from Peter, @tbekolay have you looked into running this stuff with TravisCI? It seems to be non-trivial.

**Comment by tbekolay** [comment]
It definitely takes some doin'. I can help with setting up the `.travis.yml`, but [this](http://docs.travis-ci.com/user/gui-and-headless-browsers/#Using-xvfb-to-Run-Tests-That-Require-GUI-(e.g.-a-Web-browser)) seems to be a good place to start. Basically, TravisCI comes with Firefox and a way to get a virtual GUI environment, so it should be possible to test with Selenium on TravisCI (for the Firefox driver, in any case).

**Comment by Seanny123** [comment]
Yeah, I did see that, however that appears to be for FireFox only.
Additionally, for other web-browsers it seems the only way to (easily)
integrate it with TravisCI is to use a commercial server from Sauce. I
guess it's not impossible to run our own server, but this seems kind of
like the same problem we ran into when we thought about testing our builds
on Windows. Thoughts?

**Comment by tbekolay** [comment]
Sauce Labs cost $$$ so probably not worth it. Testing through TravisCI on Firefox will definitely catch many issues that we wouldn't catch otherwise; we will miss a few things here and there not testing in Chrome and Safari, but it's not worth all of the extra effort to test in Chrome for those last 5% of issues when testing in Firefox will get 95% of them.

And, maybe someone will end up making a free alternative for that, like they did on Windows with AppVeyor.

**Comment by Seanny123** [comment]
A further detail that I'm a bit confused about is if we should be running JavaScript specific tests with Phantom.js and some JavaScript testing framework. I have zero experience in the domain and my Googling isn't telling me much, other than Phantom.js is used for basic functionality testing. Does anyone have experience on when/how to use Phantom.js?

**Comment by Seanny123** [comment]
Additionally, to get started configuring this repository for testing with TravisCI, I would need to have admin rights. @tcstewar how would you feel about temporarily giving me admin rights?

**Comment by tcstewar** [comment]
> A further detail that I'm a bit confused about is if we should be running JavaScript specific tests with Phantom.js and some JavaScript testing framework. I have zero experience in the domain and my Googling isn't telling me much, other than Phantom.js is used for basic functionality testing. Does anyone have experience on when/how to use Phantom.js?

Never used it, but my guess is that the vast majority of the stuff on the JS side of nengo_gui is interface-related, rather than basic functionality.  I could maybe see testing some bits of the DataStore class that way, but that's about all that's jumping out.  So my guess is that phantom.js isn't all that useful for our testing.  That said, I've never actually used it so if anyone has any actual experience that should override my guess here.  :)

**Comment by tcstewar** [comment]
> @tbekolay says that all you would need to do is add me to the team nengo-gui-maintainers

Sure!  You're on the list now.

**Comment by Seanny123** [comment]
Now that I've got a few tests running with the Nengo GUI, it's time to scale up. This is how I'm envisioning structuring the tests.

![image](https://cloud.githubusercontent.com/assets/1918825/8852546/dc470870-3123-11e5-933e-4650fb276795.png)

Additionally, in terms of `pytest` structure, I'd like each function to still be a test, but only a new instance of the GUI should be started if you use a new class.

Opinions?

**Comment by tbekolay** [comment]
Structure seems good, though just in terms of where the folder should go, the general one should be in `nengo_gui/tests` and the component ones should be in `nengo_gui/components/tests` and *not* in `nengo_gui/tests/components`. Every Python project I've looked at does it in the former way, so we should be consistent with that.

I'm not familiar enough with the implementation of things to comment on reusing the same Nengo GUI instance within tests in the same class... seems fine, as long as everyone's aware that that's what happens!

**Comment by jgosmann** [comment]
It might be good to have some more structure in the general tests folder. I suppose we won't just test for bugs, but make sure some general functionality not necessarily tied to one specific component works (e.g., that the border of an object stays below the mouse cursor while resizing).

**Comment by Seanny123** [comment]
@tcstewar I've decided that I want some JavaScript specific tests due to the modifications I'm making to `Nengo.DataStore` for #528 . However, I'm at a bit of a cross-roads on how to accomplish this.

One option is to run tests via command line which is fast, but it's going to require some boilerplate according to [this answer I got on StackOverflow](http://stackoverflow.com/a/32020523/1079075).

The other option is to run the tests in a headless browser, but that's [going to require some boilerplate as well](http://developer.telerik.com/featured/journey-client-side-testing-javascript/).

Which of the two evils is the lesser in your opinion?

**Comment by Seanny123** [comment]
It's apparently quite possible to run Chrome tests in TravisCI, according to this [SO question](http://stackoverflow.com/questions/19255976/how-to-make-travis-execute-angular-tests-on-chrome-please-set-env-variable-chr) and this [blog post](https://blog.andyet.com/2015/07/28/selenium-travis-webrtc/).
